src/AggregateFunctions/UniqExactSet.h

```diff
@@ -117,7 +117,8 @@ class UniqExactSet
 
     auto merge(const UniqExactSet & other, ThreadPool * thread_pool = nullptr, std::atomic<bool> * is_cancelled = nullptr)
     {
-        if (isSingleLevel() && other.isTwoLevel())
+        const bool worth_converting_to_two_level = other.isTwoLevel() || (worthConvertingToTwoLevel(std::max(size(), other.size())));
```

Comment:
we used to have a similar logic here:

https://github.com/ClickHouse/ClickHouse/blob/f0eae2c07efe13c646bf4cd5a030b646606ff3b1/src/AggregateFunctions/UniqExactSet.h#L97-L102

but it was based on an ad-hoc threshold and didn't account for the current size of our set (it may grow because we merged other sets into it).